src/include/storage/garbage_collector.h

```diff
@@ -51,19 +56,151 @@ class GarbageCollector {
+   * compacted undo record
+   * @param start_record the undo record which marks the beginning of compacted interval
+   * @param end_record the undo record which marks the end of compacted interval
+   * @return the compacted undo record
+   */
+  UndoRecord *CreateUndoRecord(UndoRecord *start_record, UndoRecord *end_record);
+
+  /**
+   * Given the time stamp and table slot after first GC pass, create a undo record which can accommodate all the columns
+   * found in the combined delta in the first GC pass
+   * @param timestamp the timestamp which will be allotted to the compacted undo record
+   * @param slot the tuple slot corresponding to the tuple associated with the compacted undo record
+   * @param table the table corresponding to the tuple associated with the compacted undo record
+   * @return the blank undo record which is the placeholder for the compacted undo record
+   */
+  UndoRecord *InitializeUndoRecord(transaction::timestamp_t timestamp, TupleSlot slot, DataTable *table);
+
+  /**
+   * Given the undo record,  mark all the varlen entries in the delta to be available for deallocation later
+   * @param undo_record the undo record whose varlen entries are to be marked for deallocation
+   */
+  void MarkVarlenReclaimable(UndoRecord *undo_record);
+
+  /**
+   * Given the undo buffer, deallocate all the varlen entries contained in all the undo records in the undo buffer
+   * @param undo_buffer the undo buffer whose varlen entries are to deallocated
+   */
+  void DeallocateVarlen(UndoBuffer *undo_buffer);
+
+  /**
+   * Given the compacted undo record, duplicate all the varlens associated with it
+   * @param undo_record the compacted undo record
+   */
+  void CopyVarlen(UndoRecord *undo_record);
 
-  transaction::TransactionManager *const txn_manager_;
+  // reference to the transaction manager class object
+  transaction::TransactionManager *txn_manager_;
   // timestamp of the last time GC unlinked anything. We need this to know when unlinked versions are safe to deallocate
   transaction::timestamp_t last_unlinked_;
   // queue of txns that have been unlinked, and should possible be deleted on next GC run
   transaction::TransactionQueue txns_to_deallocate_;
   // queue of txns that need to be unlinked
   transaction::TransactionQueue txns_to_unlink_;
+  // Undo buffer to hold compacted undo records
+  storage::UndoBuffer *delta_record_compaction_buffer_;
+  // Variable to mark that undo buffer to hold compacted undo records is empty so that it can be deallocated without
+  // unlinking
+  bool compaction_buffer_empty;
+  // queue of undo buffers containing compacted undo records which are pending unlinking
+  std::forward_list<storage::UndoBuffer *> buffers_to_unlink_;
+  // queue of undo buffers containing compacted undo records which ahve been unlinked abd are pending deallocation
+  std::forward_list<storage::UndoBuffer *> buffers_to_deallocate_;
```

Comment:
According to my understanding, the memory of compacted records is allocaed by interval GC, which needs to be reclaimed through buffers_to_unlink/buffers_to_deallocate_. If I am wrong, please correct me. 
The difference between undorecords in buffers_to_unlink/buffers_to_deallocate_ and undorecords in txns_to_unlink_/txns_to_deallocate_  is a little confusing, it is helpful if you could provide detailed explaination in design document or comments.